File: golang/数组/长度最小的子数组.go
package main

func minSubArrayLen(target int, nums []int) int {
	n := len(nums)
	left, right := 0, 0
	sum := 0
	minLen := n + 1

	for right < n {
		sum += nums[right]
		for sum >= target {
			if right-left+1 < minLen {
				minLen = right - left + 1
			}

			sum -= nums[left]
			left++
		}

		right++
	}

	if minLen == n+1 {
		return 0
	}

	return minLen
}

File: python/数组/长度最小的子数组.py
def minSubArrayLen(target, nums):
    # 滑动窗口法
    n = len(nums)
    left = 0
    right = 0
    min_len = n + 1
    sum = 0

    while right < n:
        sum += nums[right]

        while sum >= target:
            min_len = min(min_len, right - left + 1)
            sum -= nums[left]
            left += 1

        right += 1

    return min_len if min_len <= n else 0


if __name__ == "__main__":
    target = 7
    nums = [2, 3, 1, 2, 4, 3]
    min_len = minSubArrayLen(target, nums)
    print(min_len)
